### PR TITLE
Bug 1690088: Use explicitly passed-in buildinfo, if provided

### PR DESCRIPTION
**Pull request by mdboom**
Depends on https://github.com/mozilla/glean_parser/pull/276.

The sample application demonstrates the straightforward changes that would be required of Fenix to use this.  This shouldn't be a breaking change, as the new parameter is optional.  Ideally, though, once we've confirmed this is a better approach, we would make it required and then remove the package manager code path.

### CODE REVIEW

**Review by badboy — COMMENTED** [review]
(please re-request review when turning this from draft into full PR)

Just as straight forward as the glean_parser changes.

I'd like @Dexterp37's eyes on this before we land it.
Can we also do a quick check in Fenix that what our `buildInfo` reports and what we currently report is the same? I could do that tomorrow on my local dev build by adding some logging.

**Comment by mdboom** [comment]
@badboy: This is using a real glean_parser release and passing CI.  I agree @Dexterp37 should have a look.  I have not integrated this into Fenix locally to double check that the data matches.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt
Comment (Dexterp37) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt:147]: If we wanted to get fancy, we could probably create an overload for `Glean.initialize` without `buildInfo` (i.e. the current signature) and mark it as [deprecated](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-deprecated/) (which talks nicely to the compiler and Android Studio!).

It's a super nit, not very important, though.

File: glean-core/android/src/test/java/mozilla/telemetry/glean/GleanTest.kt
Comment (Dexterp37) [inline glean-core/android/src/test/java/mozilla/telemetry/glean/GleanTest.kt:819]: Can we also add a similar test to the [Java compatibility test](https://searchfox.org/glean/source/glean-core/android/src/test/java/mozilla/telemetry/glean/GleanFromJavaTest.java)?

File: docs/user/general-api.md
Comment (Dexterp37) [inline docs/user/general-api.md:66]: I wonder if future versions of the language bindings could even do without this, with Glean automatically fetching the generated code.

**Review by Dexterp37 — APPROVED** [review]
r+wc! This looks nice, I agree with @badboy , let's make sure this works in Fenix before triggering the domino-updates-effect!

File: docs/user/general-api.md
Comment (badboy) [reply docs/user/general-api.md:66]: Not really? We need to generate this in the final application because only there the build info is valid (or even existent?)
The only way we get to that data is if someone gives it to us (this is the same problem as with `registerPings`: We don't know about them unless they get passed in)

**Comment by badboy** [comment]
Oh one more thing I just realized: Is the build info stuff valid for libraries? If we generate code that doesn't work for libraries that would be bad.

**Comment by mdboom** [comment]
> Oh one more thing I just realized: Is the build info stuff valid for libraries? If we generate code that doesn't work for libraries that would be bad.

From the docs, I think it *should*, but we will know when we integrate this in android-components.  This definitely wouldn't work if the library is built with something other than Gradle (since the `BuildConfig` object this uses is injected by Gradle).  But Glean isn't really supported on non-Gradle build systems at the moment anyway...

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt
Comment (mdboom) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt:147]: This is a good idea.  Might as well...

File: glean-core/android/src/test/java/mozilla/telemetry/glean/GleanTest.kt
Comment (mdboom) [reply glean-core/android/src/test/java/mozilla/telemetry/glean/GleanTest.kt:819]: Yep.  Good idea.

Comment (mdboom) [reply glean-core/android/src/test/java/mozilla/telemetry/glean/GleanTest.kt:828]: I struggled to use the Glean testing API from those Java tests, and it doesn't seem like anything else there does that already.  I filed [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1692506) for that, and maybe we don't need to tackle that here?

Comment (Dexterp37) [reply glean-core/android/src/test/java/mozilla/telemetry/glean/GleanTest.kt:828]: Oh, interesting. Sure that's fine!

File: glean-core/android/src/test/java/mozilla/telemetry/glean/GleanFromJavaTest.java
Comment (Dexterp37) [inline glean-core/android/src/test/java/mozilla/telemetry/glean/GleanFromJavaTest.java:90]: nit: can you reference the bug you just filed about the testing API in java, within this comment?

**Review by Dexterp37 — APPROVED** [review]

File: glean-core/android/src/test/java/mozilla/telemetry/glean/GleanFromJavaTest.java
Comment (mdboom) [reply glean-core/android/src/test/java/mozilla/telemetry/glean/GleanFromJavaTest.java:90]: Done

File: CHANGELOG.md
Comment (badboy) [reply CHANGELOG.md:6]: Can we call out that this makes _not passing it_ deprecated?

**Review by badboy — APPROVED** [review]
Looks good.

Can we maybe squash this into 2 reasonable commits?
1. Upgrade glean_parser
2. The rest of this PR

**Pull request merged**
